python/pdo/common/key_value.py

```diff
@@ -0,0 +1,326 @@
+        :return string: block data
+        """
+
+        block_hash = self.input_encoding_conversion[input_encoding](block_id)
+
+        raw_block_data = kvs.block_store_get(block_hash)
+        block_data = array('B', raw_block_data).tobytes()
+
+        return block_data
+
+    # -----------------------------------------------------------------
+    def __get_block_iterator__(self, block_ids, input_encoding='b64') :
+        """Return the data for blocks given a list of block identifiers
+
+        :param block_ids list of string: block identifiers
+        :param input_encoding string: encoding to use for block identifiers, raw/b64
+        :return string: block data
+        """
+
+        for block_id in block_ids :
+            block_data = self.__get_block__(block_id, input_encoding)
+            if block_data is None :
+                raise Exception('unable to locate required block; {}'.format(block_id))
+            yield block_data
+
+    # -----------------------------------------------------------------
+    def __store_block__(self, block_data, duration=60, input_encoding='str') :
+        """
+        """
+
+        block_data = self.input_encoding_conversion[input_encoding](block_data)
+        block_hash = hashlib.sha256(block_data).digest()
+        block_hash = pcrypto.string_to_byte_array(block_hash)
+
+        kvs.block_store_put(block_hash, block_data)
+
+    # -----------------------------------------------------------------
+    def sync_to_block_store(self, dst_block_store, **kwargs) :
+        if self.__handle__ is not None :
+            raise Exception("key value store must be closed to sync")
+
+        default_minimum_duration = pconfig.shared_configuration(['Replication', 'MinimumDuration'], 5)
+        minimum_duration = kwargs.get('minimum_duration', default_minimum_duration)
+
+        default_duration = pconfig.shared_configuration(['Replication', 'Duration'], 60)
+        duration = kwargs.get('duration', default_duration)
+
+        try :
+            root_block = (self.__get_block__(self.hash_identity, input_encoding='b64'))
+        except Exception :
+            logger.exception('failed to get root block from local kv store')
+            raise
+
+        try :
+            root_block = root_block.decode('utf8')
+        except AttributeError :
+            pass
+
+        root_block = root_block.rstrip('\0')
+        root_block_json = json.loads(root_block)
+
+        block_ids = [self.hash_identity] + root_block_json['BlockIds']
+
+        # check to see which blocks need to be pushed
+        blocks_to_extend = []
+        blocks_to_push = []
+        block_status_list = dst_block_store.check_blocks(block_ids)
+
+        for block_status in block_status_list :
+            # if the size is 0 then the block is unknown to the storage service
+            if block_status['size'] == 0 :
+                blocks_to_push.append(block_status['block_id'])
```

Comment:
(minor) for clarity, if this is needed elsewhere, it would be better to have this check in a separate function